palace/drivers/basesolver.cpp

```diff
@@ -84,6 +89,282 @@ BaseSolver::BaseSolver(const IoData &iodata, bool root, int size, int num_thread
+          mesh::DimensionalizeMesh(smesh, iodata.GetMeshScaleFactor());
+          smesh.Mesh::Print(serial);  // Do not need to nondimensionalize the temporary
+        }
+        Mpi::Barrier(comm);
+      }
+      else
+      {
+        auto smesh = std::make_unique<mfem::Mesh>(mesh->GetSerialMesh(0));
+        BlockTimer bt(Timer::IO);
+        if (Mpi::Rank(comm) == 0)
+        {
+          std::ofstream serial(serial_mesh_filename);
+          serial.precision(mesh::MSH_FLT_PRECISION);
+          mesh::DimensionalizeMesh(*smesh, iodata.GetMeshScaleFactor());
+          smesh->Print(serial);  // Do not need to nondimensionalize the temporary
+        }
+      }
+    }
+    mesh->ExchangeFaceNbrData();
+  }
+}
+
+}  // namespace
+
+void BaseSolver::SolveEstimateMarkRefine(
+    std::vector<std::unique_ptr<mfem::ParMesh>> &mesh) const
+{
+  const auto &param = iodata.model.refinement;
+  auto comm = mesh.back()->GetComm();
+  // Helper to save off postprocess data.
+  auto SavePostProcess = [&](int iter)
+  {
+    Mpi::Barrier(comm);
+    BlockTimer bt_io(Timer::IO);
+    if (Mpi::Root(comm) && param.save_adapt_iterations)
+    {
+      // Create a subfolder for the results of this adaptation.
+      namespace fs = std::filesystem;
+      const auto out_dir = fs::path(IterationPostDir(iter));
+      constexpr auto options =
+          fs::copy_options::recursive | fs::copy_options::overwrite_existing;
+      const fs::path root_dir(post_dir);
+      for (const auto &f : fs::directory_iterator(root_dir))
+      {
+        if (f.is_regular_file())
+        {
+          fs::copy(f, out_dir / fs::path(f).filename(), options);
+        }
+      }
+      if (fs::exists(root_dir / "paraview"))
+      {
+        fs::copy(root_dir / "paraview", out_dir / "paraview", options);
+      }
+    }
+    Mpi::Barrier(comm);
+  };
+
+  const bool use_amr = param.adapt_max_its > 0;
+  const bool use_coarsening = mesh.back()->Nonconforming() && param.use_coarsening;
+  if (use_amr && mesh.size() > 1)
+  {
+    Mpi::Print("{}\n", "Flattening mesh sequence: AMR will start from the final mesh in "
+                       "the refinement sequence.");
+    mesh.erase(mesh.begin(), mesh.end() - 1);
+    constexpr bool refine = true, fix_orientation = true;
+    mesh.back()->Finalize(refine, fix_orientation);
+  }
+
+  int iter = 0;
+  auto indicators_and_ntdof = Solve(mesh);
+  const auto &indicators = indicators_and_ntdof.first;
+  const auto &ntdof = indicators_and_ntdof.second;
+  if (use_amr)
+  {
+    const auto is_transient = dynamic_cast<const TransientSolver *>(this) != nullptr;
+    if (is_transient)
+    {
+      MFEM_ABORT("AMR is not currently supported for transient simulations!");
+    }
+    else
+    {
+      Mpi::Print("\nAdaptive Mesh Refinement Parameters:\n");
+      Mpi::Print("MaxIter: {}, Tolerance: {:.3e}\n\n", param.adapt_max_its,
+                 param.adapt_tolerance);
+      SavePostProcess(iter);  // Save the initial solution
```

Comment:
As discussed offline, this can become the first step of the `while` loop below so that the final solution is not duplicated on disk.